File: solutions/C++/n_queens_ii_backtracking.cpp
#include <vector>
using namespace std;

int totalNQueens(int n) {
    vector<int> board(n, -1);

    function<bool(int, int)> is_safe = [&](int row, int col) {
        for (int i = 0; i < row; i++) {
            if (board[i] == col || abs(board[i] - col) == abs(i - row))
                return false;
        }
        return true;
    };

    function<int(int)> backtrack = [&](int row) -> int {
        if (row == n) return 1;
        int count = 0;
        for (int col = 0; col < n; col++) {
            if (is_safe(row, col)) {
                board[row] = col;
                count += backtrack(row + 1);
            }
        }
        return count;
    };

    return backtrack(0);
}

File: solutions/Rust/n_queens_ii_backtracking.rs
pub fn total_n_queens(n: i32) -> i32 {
    let n = n as usize;
    let mut board = vec![-1; n];

    fn is_safe(board: &[i32], row: usize, col: i32) -> bool {
        for i in 0..row {
            if board[i] == col || (board[i] - col).abs() == (i as i32 - row as i32).abs() {
                return false;
            }
        }
        true
    }

    fn backtrack(board: &mut Vec<i32>, row: usize, n: usize) -> i32 {
        if row == n { return 1; }
        let mut count = 0;
        for col in 0..n {
            if is_safe(board, row, col as i32) {
                board[row] = col as i32;
                count += backtrack(board, row + 1, n);
            }
        }
        count
    }

    backtrack(&mut board, 0, n) as i32
}
